### PR TITLE
Namecheap provider

### PR DESCRIPTION
**Pull request by HX-Rd**
Hi @jen20 
I fixed issues mentioned in #5434 
Also changed some variable names so they are more in line with the DNSimple and DNSMadeEasy providers. Also added a calculated output parameter "hostname" that is the FQDN of the record.

### CODE REVIEW

**Comment by stack72** [comment]
Hi @HX-Rd 

Can #5434 be closed in favour of this PR?

**Comment by HX-Rd** [comment]
Hi @stack72, yeah #5434 can be closed, will do so now

**Comment by HX-Rd** [comment]
Hi guys. Was wondering about the status of this PR. Is there interest in merging this ( in which case I can update the PR to resolve the conflicts ).

**Comment by stack72** [comment]
Hi @HX-Rd 

YEs, please do rebase the PR and we can take a look at this

Paul

**Comment by HX-Rd** [comment]
One thing I want to address to whom ever that reviews this commit.
The namecheap api only allows to set the DNS records for the entire host at once ( cant add one record, or delete one at the time ). The way I addressed this is that every time you add a record, I first get the list of DNS records already on the host, add the new record to it and send in the new list. Same goes for delete. This approach can cause a raise condition ( adding two records, getting the same list both times, only the last one gets in ) so that is the reason for the mutex in the resource_namecheap_record. If there are any questions just fire them my way.
Thanks!

**Comment by jen20** [comment]
Hi @HX-Rd! Now we have 0.7 out the door, we'd like to circle back around to this pull request and get it merged in. Since it's been unreviewed for a while, we'll probably pull it into a tree here and then fix anything which needs doing prior to merge if that is ok with you?

**Comment by HX-Rd** [comment]
Hi @jen20, sounds good to me! Just let me know if there is some thing I can help out with

**Comment by choliver** [comment]
Hey @jen20 - just curious what the status of this is?  (We're about to start scoping out Terraform here, and Namecheap support would be great!)

File: builtin/providers/namecheap/config.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+type Config struct {
+	username    string
+	api_user    string
+	token       string
+	ip          string
+	use_sandbox bool
+}
+
+// Client() returns a new client for accessing namne cheap.
```
Comment (danielhoherd) [inline builtin/providers/namecheap/config.go]: s/namne cheap/Namecheap/

**Comment by chrisfu** [comment]
ping @HX-Rd 

Noticed you've been active again recently. :) Any chance of resolving the conflicts above and typo that came up in the review?

Hopefully there shouldn't be much reason now to not get this PR merged soon.

**Comment by HX-Rd** [comment]
Sure thing @chrisfu , I will rebase and fix the typo once i get home from work.

**Comment by radeksimko** [comment]
Hi @HX-Rd 
sorry to keep you waiting for so long here. Let me give you a brief status update.

We have been working in the past weeks to setup a NameCheap account in which we can run acceptance tests. This is now ready to be used, but we found that NameCheap requires us to whitelist IPs for any API clients. This doesn't work well with our test environment which launches EC2 instances on demand with random IPs from the Amazon's IP pool (that is much more than 30 IPs - limit in Namecheap). We're looking into solving this by allocating EIPs, so it's not an unsolvable problem, but it will delay shipping this provider as we have to modify our infrastructure to accommodate for this specific requirement.

Secondly I managed to run both acceptance tests manually and these have passed 👍 but I noticed there's a forgotten `Print` in the vendored library which (a) prints out whole body including the token which is a security concern and (b) clutters the test output. Would you mind removing it?
https://github.com/HX-Rd/namecheap/blob/master/api.go#L72

Thirdly to reduce the LOC here in this PR would you mind submitting the `/vendor` changes in a separate PR? I'm happy merge such PR straight away and let you rebase here.

Thanks.

**Comment by HX-Rd** [comment]
Hi @radeksimko 
I removed the print statement and ill create a new pull request once I get home from work.
Thanks!

**Comment by thomasbiddle** [comment]
Great work! Looks like you authored the Go package for this as well; so my comment here applies to that as much as this :-)

Definitely don't want it to hinder the progress of this being merged in, as the initial merge is most important, but adding `namecheap.domains.dns.setCustom` API call would be wonderful for setting custom name servers (Use-case: Configuring AWS Route53, and then updating my domain's DNS to there).

Looks like there are plenty other methods that could be added as well, looking forward to seeing this get merged in.

**Comment by HX-Rd** [comment]
Hi @thomasbiddle 
Yeah sounds good, lets revisit this once this gets merged in. Thanks for the feedback!

**Comment by helgihaf** [comment]
@radeksimko Any updates?

**Comment by tk8817** [comment]
Would love to know if this is blocked by anything in particular. Happy to lend a hand for both code & documentation. Great work by @HX-Rd so far to get this to where it is!

**Comment by aleclarson** [comment]
I'm just getting started with Terraform, and was really hoping Namecheap was supported. It's great to see there's at least this PR with plenty of activity. Excited to see this merged so I can stop updating my IPv4/IPv6 whenever I relaunch an instance. Cheers! 🥂

**Comment by lcacciagioni** [comment]
SLA on this?

**Comment by ryanohoro** [comment]
Bumping because official support for Namecheap would be very helpful. Currently limping along with snippets from other projects.

**Comment by OranguTech** [comment]
Bump.

**Comment by adamdecaf** [comment]
I'm gonna email (support@namecheap.com) about them adopting this under a new repo. 

https://github.com/namecheap

**Comment by HX-Rd** [comment]
Sure thing, my understanding was that the problem with getting this merged was due to the fact that "NameCheap requires us to whitelist IPs for any API clients". But if the quality of this PR is the reason for this not being merged then i'm happy to close it.

**Comment by adamdecaf** [comment]
My email was forwarded to a manager and I haven't heard anything back yet.

**Comment by adamdecaf** [comment]
Namecheap isn't planning on supporting this. They seem a bit confused and thought I was selling them "Terraform" anyway.. I'll try and carve out some time this week to get a proper plugin working.

**Comment by HX-Rd** [comment]
Great, since this pull request is basically deprecated anyway ( providers are no longer stored in the builtin directory like they were when this was written like two years ago ) i'm closing this. @adamdecaf so you are going to put in a new pull request for this? I would check with @radeksimko about the IP pools if that is sorted out

**Comment by adamdecaf** [comment]
@HX-Rd I created https://github.com/adamdecaf/terraform-provider-namecheap

Would you mind just PR'ing the code you have over there? Doesn't have to compile. I'll take it and cleanup, but I'd rather not copy/paste your code.

**Comment by HX-Rd** [comment]
@adamdecaf Sure thing, I'll try to get it in tonight or to morrow at the latest. Thanks for the consideration, really appreciate it.

**Comment by Dzhuneyt** [comment]
@HX-Rd @adamdecaf any movement on this? Do you need any help?

**Pull request closed without merging**